data/transactions/logic/frames_test.go

```diff
@@ -0,0 +1,443 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"testing"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+const frameNonsense = `
+  return						// proto subs must appear in deadcode
+ double:
+  proto 1 2
+  pushn 1						// one return value
+  frame_dig -1
+  int 2
+  *
+  frame_bury 0
+  retsub
+  pushint 2
+  popn 1
+  dupn 4
+  bury 9
+`
+
+const frameCompiled = "43f001024501f1ff240bf20089810246014704f309"
+
+func TestPushPopN(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// These two are dumbs uses of pushn, and should perhaps be banned
+	testAccepts(t, "pushn 0; int 1", fpVersion)
+	testAccepts(t, "pushn 1; !", fpVersion)
+
+	// These two are equally dumbs uses of popn, and should perhaps be banned
+	testAccepts(t, "int 1; popn 0", fpVersion)
+	testAccepts(t, "int 1; dup; popn 1;", fpVersion)
+
+	testAccepts(t, "pushn 2; pop; !", fpVersion)
+	testAccepts(t, "pushn 3; !; assert; !; assert; !", fpVersion)
+	testPanics(t, "pushn 2", fpVersion)
+
+	testAccepts(t, "int 1; int 1; int 1; popn 2", fpVersion)
+	testAccepts(t, "int 1; int 0; popn 1", fpVersion)
+	testPanics(t, "int 1; int 0; popn 2", fpVersion)
+	testProg(t, "int 1; int 0; popn 3", LogicVersion, Expect{1, "popn 3 expects 3..."})
+	testPanics(t, notrack("int 1; int 0; popn 3"), fpVersion)
+
+	testAccepts(t, `pushn 250; pushn 250; pushn 250; pushn 250;
+                    popn 250;  popn 250;  popn 250;  popn 249; !`,
+		fpVersion)
+	testPanics(t, `pushn 250; pushn 250; pushn 250; pushn 251
+                   popn 250;  popn 250;  popn 250;  popn 250; !`,
+		fpVersion)
+}
+
+func TestPushPopNTyping(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	testProg(t, "pushn 2; +", LogicVersion)
+	testProg(t, "pushn 2; concat", LogicVersion, Expect{1, "...wanted type []byte..."})
+
+	testProg(t, "popn 1", LogicVersion, Expect{1, "...expects 1 stack argument..."})
+}
+
+func TestSimpleFrame(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	testAccepts(t, `
```

Comment:
This test is failing for me. It gets to [this point of execution](https://github.com/algorand/go-algorand/blob/20eccb20cf01449461a7a6009a427893171911a7/data/transactions/logic/assembler.go#L1141) an then causes "index out of range [0] with length 0" panic

Reply:
Indeed, I was waiting for switch to be merged, so I could merge it into this branch before digging in. It's in the stack tracking code, I think maybe because of a `bury 0` or something like that.

Reply:
fixed up in the latest push.